File: windows/BlueLink.App/Protocol/DeviceNameGreeting.cs
using System.Buffers.Binary;
using System.Text;

namespace BlueLink.Protocol;

/// <summary>Optional display metadata after the unchanged eight-byte authenticated greeting.</summary>
public static class DeviceNameGreeting
{
    private static readonly UTF8Encoding Utf8 = new(false, true);
    public static byte[] Encode(string? name)
    {
        var clean = Normalize(name);
        var greeting = ProtocolGreeting.Current.Encode();
        if (clean.Length == 0) return greeting;
        var text = Utf8.GetBytes(clean);
        var value = new byte[14 + text.Length];
        greeting.CopyTo(value, 0);
        "BLDN"u8.CopyTo(value.AsSpan(8));
        BinaryPrimitives.WriteUInt16BigEndian(value.AsSpan(12), (ushort)text.Length);
        text.CopyTo(value, 14);
        return value;
    }

    public static string Decode(ReadOnlySpan<byte> value)
    {
        if (value.Length < 14 || !value.Slice(8, 4).SequenceEqual("BLDN"u8)) return "";
        var length = BinaryPrimitives.ReadUInt16BigEndian(value[12..14]);
        if (length > 384 || value.Length != 14 + length) return "";
        try { return Normalize(Utf8.GetString(value[14..])); }
        catch (DecoderFallbackException) { return ""; }
    }

    private static string Normalize(string? name)
    {
        var value = name?.Trim() ?? "";
        return value.Length <= 128 && !value.Any(char.IsControl) && Encoding.UTF8.GetByteCount(value) <= 384 ? value : "";
    }
}

File: protocol-core/src/main/java/com/bluelink/core/DeviceNameGreeting.java
package com.bluelink.core;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.nio.charset.CodingErrorAction;

/** Optional display metadata; old peers continue decoding the unchanged first eight bytes. */
public final class DeviceNameGreeting {
    private DeviceNameGreeting() {}

    public static byte[] encode(String name) {
        String clean = normalize(name);
        byte[] greeting = ProtocolGreeting.current().encode();
        if (clean.isEmpty()) return greeting;
        byte[] text = clean.getBytes(StandardCharsets.UTF_8);
        return ByteBuffer.allocate(14 + text.length).put(greeting).put(new byte[]{'B','L','D','N'})
                .putShort((short) text.length).put(text).array();
    }

    public static String decode(byte[] value) {
        if (value.length < 14 || value[8] != 'B' || value[9] != 'L' || value[10] != 'D' || value[11] != 'N') return "";
        int length = Short.toUnsignedInt(ByteBuffer.wrap(value, 12, 2).getShort());
        if (length > 384 || value.length != 14 + length) return "";
        try {
            return normalize(StandardCharsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT).decode(ByteBuffer.wrap(value, 14, length)).toString());
        } catch (java.nio.charset.CharacterCodingException ignored) { return ""; }
    }

    private static String normalize(String name) {
        String value = name == null ? "" : name.trim();
        return value.length() <= 128 && value.chars().noneMatch(Character::isISOControl)
                && value.getBytes(StandardCharsets.UTF_8).length <= 384 ? value : "";
    }
}
